tools/attach_logs_in_jira.py

```diff
@@ -0,0 +1,116 @@
+
+from consts import JIRA_PROJECT, JIRA_SUMMARY, JIRA_SERVER
+
+JIRA_ATTACHMENT_MAX_SIZE = 50 * 2**20  # 50MiB
+
+log = logging.getLogger(__name__)
+stream_handler = logging.StreamHandler(sys.stderr)
+stream_handler.setFormatter(logging.Formatter("%(asctime)s - %(name)s - %(levelname)s - %(message)s"))
+log.addHandler(stream_handler)
+
+
+def _get_or_create_jira_ticket(jira_client: jira.JIRA, summary: str) -> jira.Issue:
+    matching_issues = jira_client.search_issues(
+        f'project = "{JIRA_PROJECT}" AND summary ~ "{summary}"', fields=["attachment"]
+    )
+
+    if len(matching_issues) == 0:
+        log.info("Creating a placeholder issue for %s", summary)
+        return jira_client.create_issue(
+            project=JIRA_PROJECT,
+            summary=summary,
+            description=(
+                "Logs are still being uploaded.\nTicket description and signatures will get processed shortly..."
+            ),
+        )
+
+    if len(matching_issues) == 1:
+        log.debug("Using pre-existing issue %s", matching_issues[0])
+        return matching_issues[0]
+
+    raise RuntimeError(
+        f"Found more than one matching issue for '{summary}': "
+        f"{', '.join(issue.key for issue in matching_issues)}"
+    )
+
+
+def _archive_dir(directory: Path) -> Path:
+    archived_file = directory.with_suffix(".tar.gz")
+    log.info("Archiving directory %s to %s", directory, archived_file)
+    with tarfile.open(archived_file, "w:gz") as tar:
+        tar.add(directory, directory.name)
+
+    return archived_file
+
+
+def _split_file(path: Path, chunk_size: int) -> None:
+    log.info("Splitting %s to several files", path)
+    with path.open("rb") as whole_file:
+        read_chunk = partial(whole_file.read, chunk_size)
+        for i, chunk in enumerate(iter(read_chunk, b"")):
+            with open(f"{path}-part-{i}", "wb") as part_file:
+                part_file.write(chunk)
+
+
+def attach_logs_in_jira(source_dir: Path, jira_access_token: str):
+    jira_client = jira.JIRA(JIRA_SERVER, token_auth=jira_access_token, validate=True)
+
+    for cluster_logs_dir in source_dir.iterdir():
+        issue = _get_or_create_jira_ticket(jira_client, JIRA_SUMMARY.format(failure_id=cluster_logs_dir.name))
+
+        for artifact in cluster_logs_dir.iterdir():
+            if artifact.is_dir():
+                artifact = _archive_dir(artifact)
+
+            if artifact.stat().st_size > JIRA_ATTACHMENT_MAX_SIZE:
+                _split_file(artifact, chunk_size=JIRA_ATTACHMENT_MAX_SIZE)
+
+        for artifact in cluster_logs_dir.iterdir():
+            if artifact.is_dir() or artifact.stat().st_size > JIRA_ATTACHMENT_MAX_SIZE:
+                log.debug("Skip %s as it had been replaced with other files", artifact)
+                continue
+
+            if artifact.name in (attachment.filename for attachment in issue.fields.attachment):
+                log.debug("File %s already exists at %s. Skipping upload", artifact, issue)
```

Comment:
Do we have cases where cluster logs are updated and we would need to upload an updated archive?

Reply:
Good question. Maybe it's especially relevant for the events
I think I'll add another check of the size to understand if we need to upload updated files (no way to get a hash value like md5 as far as I can see, so it's the closets we can do apart from downloading the files from jira and do the hash calculation)

Reply:
or embed the checksum in archive's name, if size check is easy, it could be good enough

Reply:
to keep it simple (and also file names shorter) I'll do the size check
either way it should be a low risk with low possibility